staging/src/k8s.io/client-go/dynamic/fake/simple_test.go

```diff
@@ -64,3 +90,144 @@ func TestList(t *testing.T) {
+		return nil
+	}
+	if !equality.Semantic.DeepEqual(result, tc.PatchedObject) {
+		return fmt.Errorf("unexpected diff in received object: %s", diff.ObjectGoPrintDiff(tc.PatchedObject, result))
+	}
+	return nil
+}
+
+func TestPatch(t *testing.T) {
+	testCases := []PatchTestCase{
+		{
+			Name:       "jsonpatch fails with merge type",
+			Object:     newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+			PatchType:  types.StrategicMergePatchType,
+			PatchBytes: []byte(`[]`),
+			WantErrMsg: "invalid JSON document",
+		}, {
+			Name:      "jsonpatch works with empty patch",
+			Object:    newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+			PatchType: types.JSONPatchType,
+			// No-op
+			PatchBytes:    []byte(`[]`),
+			WantErr:       false,
+			PatchedObject: newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+		}, {
+			Name:      "jsonpatch works with simple change patch",
+			Object:    newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+			PatchType: types.JSONPatchType,
+			// change spec.foo from bar to foobar
+			PatchBytes:    []byte(`[{"op": "replace", "path": "/spec/foo", "value": "foobar"}]`),
+			WantErr:       false,
+			PatchedObject: newUnstructuredWithSpec(map[string]interface{}{"foo": "foobar"}),
+		}, {
+			Name:      "jsonpatch works with simple addition",
+			Object:    newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+			PatchType: types.JSONPatchType,
+			// add spec.newvalue = dummy
+			PatchBytes:    []byte(`[{"op": "add", "path": "/spec/newvalue", "value": "dummy"}]`),
+			WantErr:       false,
+			PatchedObject: newUnstructuredWithSpec(map[string]interface{}{"foo": "bar", "newvalue": "dummy"}),
+		}, {
+			Name:      "jsonpatch works with simple deletion",
+			Object:    newUnstructuredWithSpec(map[string]interface{}{"foo": "bar", "toremove": "shouldnotbehere"}),
+			PatchType: types.JSONPatchType,
+			// remove spec.newvalue = dummy
+			PatchBytes:    []byte(`[{"op": "remove", "path": "/spec/toremove"}]`),
+			WantErr:       false,
+			PatchedObject: newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+		}, {
+			Name:      "strategic merge patch fails with JSONPatch",
+			Object:    newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+			PatchType: types.StrategicMergePatchType,
+			// add spec.newvalue = dummy
+			PatchBytes: []byte(`[{"op": "add", "path": "/spec/newvalue", "value": "dummy"}]`),
+			WantErrMsg: "invalid JSON document",
+		}, {
+			Name:          "strategic merge is no-op with no changes",
+			Object:        newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+			PatchType:     types.StrategicMergePatchType,
+			PatchBytes:    []byte(`{}`),
+			WantErr:       false,
+			PatchedObject: newUnstructuredWithSpec(map[string]interface{}{"foo": "bar"}),
+		}, {
+			Name:          "strategic merge change",
+			Object:        newUnstructured(testAPIVersion, testKind, testNamespace, testName),
```

Comment:
Yes, the dynamic client deals with unstructured objects. But the patch is applied to an object at the server side, the server doesn't care if it's sent via a dynamic client or not.

Concretely, if the dynamic client sends a strategic merge patch for a custom resource (which is implemented as `Unstructured`), it will receive an error. (see https://github.com/kubernetes/kubernetes/issues/58414#issuecomment-383148632 for background).

I don't want to have a test case that demonstrates wrong use case.

Reply:
I've removed those test cases and just added a test that returns a failure for the patchtype when not specified. I feel like adding those tests is something that should have been done when the code was initially checked in, so adding those tests could be done in a follow on PR. Fair?